internal/app/apid/pkg/backend/apid.go

```diff
@@ -0,0 +1,197 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package backend
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/talos-systems/grpc-proxy/proxy"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/talos-systems/talos/api/common"
+	"github.com/talos-systems/talos/pkg/constants"
+)
+
+// APID backend performs proxying to another apid instance.
+//
+// Backend authenticates itself using given grpc credentials.
+type APID struct {
+	target string
+	creds  credentials.TransportCredentials
+
+	mu   sync.Mutex
+	conn *grpc.ClientConn
+}
+
+// NewAPID creates new instance of APID backend
+func NewAPID(target string, creds credentials.TransportCredentials) (*APID, error) {
+	// perform very basic validation on target
+	if target == "" || strings.Contains(target, ":") {
+		return nil, fmt.Errorf("invalid target %q", target)
+	}
+
+	return &APID{
+		target: target,
+		creds:  creds,
+	}, nil
+}
+
+func (a *APID) String() string {
+	return a.target
+}
+
+// GetConnection returns a grpc connection to the backend.
+func (a *APID) GetConnection(ctx context.Context) (context.Context, *grpc.ClientConn, error) {
+	origMd, ok := metadata.FromIncomingContext(ctx)
+
+	md := origMd.Copy()
+	delete(md, "targets")
+	delete(md, ":authority")
+
+	if ok {
+		md.Set("proxyfrom", origMd[":authority"]...)
+	} else {
```

Comment:
Just to make sure i'm reading this right, the `else` condition would only be met if we had a context passed through that didn't include metadata?

Reply:
yep, I thought we should include something anyways. I know `:authority` should always be there, but just to be 100% sure to make proxied request.